pkg/util/ciutil/systems/systems.go

```diff
@@ -0,0 +1,119 @@
+	Buildkite                   SystemName = "Buildkite"
+	CircleCI                    SystemName = "CircleCI"
+	Codeship                    SystemName = "Codeship"
+	Drone                       SystemName = "Drone"
+
+	// GenericCI is used when a CI system in which the CLI is being run,
+	// is not recognized by it. Users can set the relevant env vars
+	// as a fallback so that the CLI would still pick-up the metadata related
+	// to their CI build.
+	GenericCI SystemName = "Generic CI"
+
+	GitHub      SystemName = "GitHub"
+	GitLab      SystemName = "GitLab"
+	GoCD        SystemName = "GoCD"
+	Hudson      SystemName = "Hudson"
+	Jenkins     SystemName = "Jenkins"
+	MagnumCI    SystemName = "Magnum CI"
+	Semaphore   SystemName = "Semaphore"
+	TaskCluster SystemName = "TaskCluster"
+	TeamCity    SystemName = "TeamCity"
+	Travis      SystemName = "Travis CI"
+)
+
+// SystemName is a recognized CI system.
+type SystemName string
+
+// System represents a CI/CD system.
+type System interface {
+	// DetectVars when called on a specific instance of a CISystem
+	// detects the env vars of the corresponding CI/CD system and
+	// returns `Vars` with those values.
+	DetectVars() Vars
+	// IsCI returns true if any of the CI systems's associated environment variables are set.
+	IsCI() bool
+}
+
+// Vars contains a set of metadata variables about a CI system.
+type Vars struct {
+	// Name is a required friendly name of the CI system.
+	Name SystemName
+	// BuildID is an optional unique identifier for the current build/job.
+	BuildID string
+	// BuildType is an optional friendly type name of the build/job type.
+	BuildType string
+	// BuildURL is an optional URL for this build/job's webpage.
+	BuildURL string
+	// SHA is the SHA hash of the code repo at which this build/job is running.
+	SHA string
+	// BranchName is the name of the feature branch currently being built.
+	BranchName string
+	// CommitMessage is the full message of the Git commit being built.
+	CommitMessage string
+	// PRNumber is the pull-request ID/number in the source control system.
+	PRNumber string
+}
+
+// BaseCISystem implements the `System` interface with default
+// implementations.
+//
+// When creating a new CI System implementation, implement the
+// DetectVars and any other function you wish to override.
+type BaseCISystem struct {
+	Name SystemName
+	// EnvVarsToDetect is an array of env vars to check if any of these env vars is set,
+	// which would indicate that the Pulumi CLI is running in that CI system's environment.
+	EnvVarsToDetect []string
+	// EnvValuesToDetect is a map of env vars and their expected values to check for,
+	// in order to see if the Pulumi CLI is running inside a certain CI system's environment.
+	EnvValuesToDetect map[string]string
+}
+
+// DetectVars for a known CI system returns an empty Vars struct.
+func (d BaseCISystem) DetectVars() Vars {
+	return Vars{}
+}
+
+// IsCI returns true if a specific env var of a CI system is set.
```

Comment:
Would we want to err on the side of _all_ `EnvVarsToDetect` and `EnvValuesToDetect` to detect are found? In practice, it looks like only one value is defined. But at least that way we can avoid false positives?

Reply:
I would not. I mean, each CI system defines specific env vars. But also, note that some env vars are legitimately blank when certain conditions are not met. For example, the CI job could be a push build and hence the PR env vars would be blank.